modules/dxp/apps/search-experiences/search-experiences-web/src/main/resources/META-INF/resources/sxp_blueprint_admin/js/edit_sxp_blueprint/query_builder_tab/SelectTypes.js

```diff
@@ -5,39 +5,183 @@
 
 import ClayButton from '@clayui/button';
 import ClayIcon from '@clayui/icon';
-import ClayTable from '@clayui/table';
-import React, {useContext} from 'react';
+import ClayList from '@clayui/list';
+import React, {useContext, useState} from 'react';
 
 import ThemeContext from '../../shared/ThemeContext';
 import SearchableTypesModal from './SearchableTypesModal';
+import SelectSubtypes from './SelectSubtypes';
+
+/**
+ * Grabs all the selected subtypes of a certain type.
+ *
+ * @param {array} selected
+ * @param {string} className
+ * @returns {array}
+ */
+const getSelectedSubtypes = (selected, className) => {
+	return selected.find((item) => item.type === className)?.subtypes;
+};
+
+/**
+ * Turns a flat list of types and subtypes into a nested array of the types
+ * with its subtypes as children. This is called when setting up the 'selected'
+ * state and allows for easier manipulation of the data.
+ *
+ * @param {*} initialSelectedTypes
+ * @param {*} subtypeClassesMap
+ * @returns {array}
+ */
+const setupSelected = (initialSelectedTypes, subtypeClassesMap = {}) => {
+	const selected = [];
+
+	initialSelectedTypes.forEach((item) => {
+		const selectedTypes = item.split('&&');
+
+		if (selectedTypes.length < 3) {
+			selected.push({subtypes: [], type: item});
+		}
+		else {
+			const typeClassName = selectedTypes[0];
+
+			const itemIndex = selected.findIndex(
+				({type}) => type === typeClassName
+			);
+
+			if (itemIndex > -1) {
+				const subtypesArray = selected[itemIndex].subtypes || [];
+
+				subtypesArray.push({
+					label: subtypeClassesMap[item] || item,
+					value: item,
+				});
+
+				selected[itemIndex] = {
+					subtypes: subtypesArray,
+					type: typeClassName,
+				};
+			}
+			else {
+				selected.push({
+					subtypes: [
+						{label: subtypeClassesMap[item] || item, value: item},
+					],
+					type: typeClassName,
+				});
+			}
+		}
+	});
+
+	return selected;
+};
+
+/**
+ * Transforms the selected types and subtypes into a flat array that
+ * the backend expects for submission.
+ *
+ * @param {*} newSelected
+ * @returns {array}
+ */
+const transformSelected = (newSelected) => {
+	const newSearchableAssetTypes = [];
+
+	newSelected.forEach(({subtypes, type}) => {
+		if (subtypes.length) {
+			newSearchableAssetTypes.push(...subtypes.map(({value}) => value));
+		}
+		else {
+			newSearchableAssetTypes.push(type);
+		}
+	});
+
+	return newSearchableAssetTypes;
+};
```

Comment:
I don't think you need to name these variables with `new*` since it's transforming everything that's selected regardless if it's "new" or not. I'd mainly use the "new" naming if you're trying to differentiate between values that are previously selected and newly selected in the same context, but this function is only dealing with 1 set of values.